x/rewards/keeper/providers.go

```diff
@@ -39,40 +51,68 @@ func (k Keeper) distributeMonthlyBonusRewards(ctx sdk.Context) {
 		k.removeAllBasePay(ctx)
 		utils.LogLavaEvent(ctx, k.Logger(ctx), types.ProvidersBonusRewardsEventName, details, "provider bonus rewards distributed successfully")
 	}()
+
+	// Get serviced CU for each provider + spec
+	specCuMap := map[string]types.SpecCuType{} // spec -> specCu
 	for _, spec := range specs {
 		// all providers basepays and the total basepay of the spec
 		basepays, totalbasepay := k.specProvidersBasePay(ctx, spec.ChainID)
-		if totalbasepay.IsZero() {
+		if len(basepays) == 0 {
 			continue
 		}
 
 		// calculate the maximum rewards for the spec
-		specTotalPayout := k.specTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
+		specTotalPayout := math.LegacyZeroDec()
+		if !totalbasepay.IsZero() {
+			specTotalPayout = k.specTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
+		}
 		// distribute the rewards to all providers
 		for _, basepay := range basepays {
-			// calculate the providers bonus base on adjusted base pay
-			reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
-			totalRewarded = totalRewarded.Add(reward)
-			if totalRewarded.GT(total) {
-				utils.LavaFormatError("provider rewards are larger than the distribution pool balance", nil,
-					utils.LogAttr("distribution_pool_balance", total.String()),
-					utils.LogAttr("provider_reward", totalRewarded.String()))
-				details["error"] = "provider rewards are larger than the distribution pool balance"
-				return
-			}
-			// now give the reward the provider contributor and delegators
-			providerAddr, err := sdk.AccAddressFromBech32(basepay.Provider)
-			if err != nil {
-				continue
-			}
-			_, _, err = k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, reward, string(types.ProviderRewardsDistributionPool), false, false, false)
-			if err != nil {
-				utils.LavaFormatError("failed to send bonus rewards to provider", err, utils.LogAttr("provider", basepay.Provider))
+			if !specTotalPayout.IsZero() {
```

Comment:
the if should come before the loop

Reply:
done. see https://github.com/lavanet/lava/commit/80e3986929c5d425d7fa24b3fb0cd674062a52cd

Reply:
this is not true. there could be a case where the specTotalPayout is zero but i still need to count iprpc cu. If the condition was before the loop, I would skip the iprpc cu count. Reverted here: https://github.com/lavanet/lava/commit/a4062fe82fd2857e9d5683481942b811982870ca